### PR TITLE
mem-hotplug: fix flaky test and small issues

### PR DESCRIPTION
**Pull request by acardace**
This test proved to be a bit flaky
https://search.ci.kubevirt.io/?search=Memory+Hotplug+A+VM+with+memory+liveUpdate+enabled+should+successfully+hotplug+memory+twice&maxAge=48h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job.

This makes sure we wait for the migration to have succeeded before inspecting the domain XML, otherwise the domain might not be running yet.

### What this PR does
Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/hotplug/memory.go [hunk 1]
```diff
@@ -290,6 +290,24 @@ var _ = Describe("[sig-compute][Serial]Memory Hotplug", decorators.SigCompute, d
 			_, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchBytes, k8smetav1.PatchOptions{})
 			Expect(err).ToNot(HaveOccurred())
 
+			By("Waiting for HotMemoryChange condition to appear")
+			Eventually(ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceMemoryChange))
+
+			By("Ensuring live-migration started")
+			var migration *v1.VirtualMachineInstanceMigration
+			Eventually(func() bool {
+				migrations, err := virtClient.VirtualMachineInstanceMigration(vm.Namespace).List(context.Background(), k8smetav1.ListOptions{})
```
Comment (fossedihelm) [inline tests/hotplug/memory.go:299]: You can filter migrations by `"kubevirt.io/vmi-name"` label with `vmi.Name` as a value, saving all the logic below.

File: tests/hotplug/memory.go [hunk 2]
```diff
@@ -341,6 +359,24 @@ var _ = Describe("[sig-compute][Serial]Memory Hotplug", decorators.SigCompute, d
 				vm, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchBytes, k8smetav1.PatchOptions{})
 				Expect(err).ToNot(HaveOccurred())
 
+				By("Waiting for HotMemoryChange condition to appear")
+				Eventually(ThisVMI(vmi), 1*time.Minute, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceMemoryChange))
+
+				By("Ensuring live-migration started")
+				var migration *v1.VirtualMachineInstanceMigration
+				Eventually(func() bool {
+					migrations, err := virtClient.VirtualMachineInstanceMigration(vm.Namespace).List(context.Background(), k8smetav1.ListOptions{})
```
Comment (fossedihelm) [inline tests/hotplug/memory.go:368]: same as above

File: tests/hotplug/memory.go (hunk 1 above)
Comment (acardace) [reply tests/hotplug/memory.go]: Done.

File: tests/hotplug/memory.go (hunk 2 above)
Comment (acardace) [reply tests/hotplug/memory.go]: Done.

File: pkg/virt-controller/watch/vm.go [hunk 3]
```diff
@@ -3316,6 +3312,17 @@ func (c *VMController) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi
...
+		vmConditions := controller.NewVirtualMachineConditionManager()
+		vmConditions.UpdateCondition(vm, &virtv1.VirtualMachineCondition{
+			Type:               virtv1.VirtualMachineRestartRequired,
+			LastTransitionTime: metav1.Now(),
+			Status:             k8score.ConditionTrue,
+			Message:            "memory updated in template spec. Memory-hotplug is not available for this VM configuration",
+		})
+		return nil
+	}
+
```
Comment (fossedihelm) [inline pkg/virt-controller/watch/vm.go]: We should cover this with a unit test too

File: tests/hotplug/memory.go [hunk 4]
```diff
@@ -434,20 +401,26 @@ var _ = Describe("[sig-compute][Serial]Memory Hotplug", decorators.SigCompute, d
...
+			LabelSelector: fmt.Sprintf("kubevirt.io/vmi-name=%s", vmi.Name),
+		}
+		migrations, err := virtClient.VirtualMachineInstanceMigration(vmi.Namespace).List(context.Background(), listOpts)
+		ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		if len(migrations.Items) > 0 {
+			migration = migrations.Items[0].DeepCopy()
+			return true
+		}
+		return false
+	}, 30*time.Second, time.Second).Should(BeTrue())
```
Comment (fossedihelm) [inline tests/hotplug/memory.go:424]: Please also provide an additional text:
```
	}, 30*time.Second, time.Second).Should(BeTrue(), "A vmim should be created")
```

File: pkg/virt-controller/watch/vm.go [hunk 5]
```diff
@@ -3316,6 +3312,17 @@ func (c *VMController) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi
 		return nil
 	}
 
+	if vmi.Spec.Domain.Memory.MaxGuest == nil {
+		vmConditions := controller.NewVirtualMachineConditionManager()
+		vmConditions.UpdateCondition(vm, &virtv1.VirtualMachineCondition{
+			Type:               virtv1.VirtualMachineRestartRequired,
+			LastTransitionTime: metav1.Now(),
+			Status:             k8score.ConditionTrue,
+			Message:            "memory updated in template spec. Memory-hotplug is not available for this VM configuration",
```
Comment (fossedihelm) [inline pkg/virt-controller/watch/vm.go]: We should cover this with a unit test too

File: pkg/virt-controller/watch/vm.go [hunk 6]
```diff
@@ -3297,17 +3297,13 @@ func (c *VMController) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi
...
-		conditionManager.HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceMemoryChange, k8score.ConditionFalse) {
-		// memory hotplug either failed or is incompatible, let's trigger the restart required condition
-		// to make the memory bump effective
-
+	if conditionManager.HasConditionWithStatus(vmi, virtv1.VirtualMachineInstanceMemoryChange, k8score.ConditionFalse) {
 		vmConditions := controller.NewVirtualMachineConditionManager()
 		vmConditions.UpdateCondition(vm, &virtv1.VirtualMachineCondition{
 			Type:               virtv1.VirtualMachineRestartRequired,
 			LastTransitionTime: metav1.Now(),
 			Status:             k8score.ConditionTrue,
```
Comment (alicefr) [inline pkg/virt-controller/watch/vm.go]: Does this still remain true?

File: pkg/virt-controller/watch/vm.go [hunk 7]
```diff
@@ -3316,6 +3312,17 @@ func (c *VMController) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi
 		return nil
 	}
 
+	if vmi.Spec.Domain.Memory.MaxGuest == nil {
+		vmConditions := controller.NewVirtualMachineConditionManager()
+		vmConditions.UpdateCondition(vm, &virtv1.VirtualMachineCondition{
+			Type:               virtv1.VirtualMachineRestartRequired,
+			LastTransitionTime: metav1.Now(),
+			Status:             k8score.ConditionTrue,
```
Comment (alicefr) [inline pkg/virt-controller/watch/vm.go]: Same question as above

**Comment by alicefr** [comment]
@acardace general question about the change condition, if it isn't possible to migrate for any reason, shouldn't we put the status to false?

File: pkg/virt-controller/watch/vm.go (hunk 6 above)
Comment (acardace) [reply pkg/virt-controller/watch/vm.go]: sorry Alice can you elaborate a bit? What does still remain true?

**Comment by acardace** [comment]
> @acardace general question about the change condition, if it isn't possible to migrate for any reason, shouldn't we put the status to false?

Which status?

File: pkg/virt-controller/watch/vm.go (hunk 6 above)
Comment (alicefr) [reply pkg/virt-controller/watch/vm.go]: If the change cannot be applied for any reason, I would put the status of the condition to false.

**Comment by alicefr** [comment]
> > @acardace general question about the change condition, if it isn't possible to migrate for any reason, shouldn't we put the status to false?
> 
> Which status?

The status of the change condition

**Review by xpivarc — COMMENTED** [review]
/approve
/hold
Few comments but otherwise looks good

File: tests/hotplug/memory.go [hunk 8]
```diff
@@ -290,6 +290,24 @@ var _ = Describe("[sig-compute][Serial]Memory Hotplug", decorators.SigCompute, d
 			_, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchBytes, k8smetav1.PatchOptions{})
 			Expect(err).ToNot(HaveOccurred())
 
+			By("Waiting for HotMemoryChange condition to appear")
```
[the remark this answers is not in the document]
Comment (acardace) [reply tests/hotplug/memory.go:282]: Sure.

File: tests/hotplug/memory.go (hunk 4 above)
Comment (acardace) [reply tests/hotplug/memory.go:424]: Done.

File: pkg/virt-controller/watch/vm.go (hunk 5 above)
Comment (acardace) [reply pkg/virt-controller/watch/vm.go]: Done

**Comment by acardace** [comment]
@xpivarc I added 3fcb188cb295378a2dbdecec9c44b756260776b6, PTAL.

**Review by fossedihelm — APPROVED** [review]
@xpivarc 
Feel free to unhold if the new changes look good to you. Thanks

**Comment by fossedihelm** [comment]
@acardace unit test failures are related

**Comment by acardace** [comment]
/unhold

@xpivarc FYI

**Comment by acardace** [comment]
/hold

Sorry @xpivarc, I forgot I added a new commit for a second :)

File: pkg/virt-controller/watch/vm.go (hunk 6 above)
Comment (xpivarc) [reply pkg/virt-controller/watch/vm.go]: We set the condition when a live update cannot be applied.

File: pkg/virt-controller/watch/vm.go [hunk 9]
```diff
@@ -3294,6 +3294,11 @@ func (c *VMController) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi
 		return nil
 	}
 
+	if !vmi.IsMigratable() {
```
Comment (xpivarc) [inline pkg/virt-controller/watch/vm.go:3297]: This should be higher in the stack

**Review by xpivarc — COMMENTED** [review]
I am pretty sure I posted this comment but I don't see it so here we go again

File: pkg/virt-controller/watch/vm.go (hunk 9 above)
Comment (acardace) [reply pkg/virt-controller/watch/vm.go:3297]: It's actually the first thing that gets checked after a change in memory is detected.

Comment (xpivarc) [reply pkg/virt-controller/watch/vm.go:3297]: I mean this is common for cpu/memory/affinity... Shouldn't it be before calling `handleMemoryHotplugRequest`

Comment (acardace) [reply pkg/virt-controller/watch/vm.go:3297]: Yes but we cannot do it right now, I was planning to work on that as part of the interface work for live update handlers. The issue right now is that you cannot tell whether a hotplug has been requested unless you call the handlers, so this check has to be done in the handlers for the time being.

Otherwise you'd set RestartRequired regardless of whether the VM owner is trying to change anything.

tldr; we have lots to do in this area.

Comment (xpivarc) [reply pkg/virt-controller/watch/vm.go:3297]: Ah, you are right... I was dreaming...

Comment (acardace) [reply pkg/virt-controller/watch/vm.go:3297]: Just looking at the future :+1:

**Comment by enp0s3** [comment]
@acardace Hey!
Will the RestartRequired condition disappear when the VM owner changes the desired configuration back to the current one AKA no restart required?
